pkg/runner/cluster_k8s.go

```diff
@@ -1187,31 +1188,40 @@ func (c *ClusterK8sRunner) GetClusterCapacity() (int64, int64, error) {
 	return allocatableCPUs, allocatableMemory, nil
 }
 
-func (c *ClusterK8sRunner) updateRunResult(template *runtime.RunParams, result *Result) error {
-	events, err := c.syncClient.FetchAllEvents(template)
+func (c *ClusterK8sRunner) collectOutcomes(ctx context.Context, result *Result, tpl *runtime.RunParams) error {
+	eventsCh, err := c.syncClient.SubscribeEvents(ctx, tpl)
 	if err != nil {
 		return err
 	}
 
-	for _, e := range events {
-		// for now we emit only outcome OK events, so no need for more checks
-		if e.SuccessEvent != nil {
-			se := e.SuccessEvent
-			o := result.Outcomes[se.TestGroupID]
-			o.Ok = o.Ok + 1
+	go func() {
+		running := true
+		for running {
+			select {
+			case <-ctx.Done():
+				running = false
+			case e := <-eventsCh:
+				// for now we emit only outcome OK events, so no need for more checks
+				if e.SuccessEvent != nil {
+					se := e.SuccessEvent
+					o := result.Outcomes[se.TestGroupID]
+					o.Ok = o.Ok + 1
+				}
+			}
 		}
-	}
 
-	result.Outcome = task.OutcomeSuccess
-	if len(result.Outcomes) == 0 {
-		result.Outcome = task.OutcomeFailure
-	}
-	for g := range result.Outcomes {
-		if result.Outcomes[g].Total != result.Outcomes[g].Ok {
+		result.Outcome = task.OutcomeSuccess
+		if len(result.Outcomes) == 0 {
 			result.Outcome = task.OutcomeFailure
-			break
 		}
-	}
+
+		for g := range result.Outcomes {
+			if result.Outcomes[g].Total != result.Outcomes[g].Ok {
```

Comment:
@hacdias I believe we have a race here - how do we know that this has completed in the main thread, so that we return `result` only after this has been run?

Reply:
Fixed on 7b7f16e. Thanks :D